### PR TITLE
Periodically save state and store in indexedDB

### LINKED ISSUE #4161: Feature Request: add crash restore
Today my Atom crashed and I realized that there is no such thing as restore your unsaved work.
Since this is more than standard nowadays - please add this!

### PR DESCRIPTION
**Pull request by kuychaco**
Fixes [atom/atom#4161](https://github.com/atom/atom/issues/4161)
- [x] separate serialization from package deactivation
- [x] extract serialization logic from unloadEditorWindow
- [x] fix package manager tests
- [x] save state asynchronously on keydown and mousedown, write tests
- [x] get rid of AtomEnvironment state instance variable
- [x] fix AtomEnvironment tests
- [x] profile serialization time for various environment states
- [x] store state in IndexedDB
- [x] add command-line flag to clear indexedDB object store

### CODE REVIEW

File: spec/atom-environment-spec.coffee [hunk 1]
```diff
@@ -174,6 +174,19 @@ describe "AtomEnvironment", ->
...
+      keypress = new KeyboardEvent('keypress')
+      atom.document.dispatchEvent(keypress)
+      advanceClock 1100
+      expect(atom.saveStateSync).toHaveBeenCalled()
+
+      mousedown = new MouseEvent('mousedown')
+      atom.document.dispatchEvent(mousedown)
+      advanceClock 1100
+      expect(atom.saveStateSync).toHaveBeenCalled()
+
```
Comment (kuychaco) [inline spec/atom-environment-spec.coffee]: Spy not working as expected and causing test to fail:

screen shot 2016-01-27 at 11 18 09 am

File: spec/package-manager-spec.coffee [hunk 2]
```diff
@@ -821,6 +819,35 @@ describe "PackageManager", ->
           expect(atom.packages.isPackageActive("package-with-missing-provided-services")).toBe true
           expect(addErrorHandler.callCount).toBe 0
 
+  describe "serialization", ->
+
```
Comment (nathansobo) [inline spec/package-manager-spec.coffee]: Super minor nitpick drive-by comment: :fire: this empty line. :smile:

**Comment by jeancroy** [comment]
Is this saved to .atom/storage ?
On the forum, when atom crash on startup, deleting the storage folder is a standard debug step.
Maybe there's an interest in discarding most of the state but restoring unsaved file content.

File: spec/atom-environment-spec.coffee (hunk 1 above)
Comment (kuychaco) [inline spec/atom-environment-spec.coffee]: Figured out this was an issue with binding `saveStateSync` before spying on it

**Comment by kuychaco** [comment]
@jeancroy thanks for your comment. To answer your question - yes, the atom environment state is saved to .atom/storage. That said, I'm going to create an unsaved text buffers store that will be located elsewhere (possibly using IndexedDB). So in the event of a crash and deletion of the storage folder, your unsaved buffers will be preserved. 

The unsaved buffer store will be a part of a separate pull request. I've updated the title of this one to be more specific as to what aspect of crash recovery will be addressed here.

**Comment by maxbrunsfeld** [comment]
This is looking good! How long does `saveStateSync` on average? Does it vary much depending on how many open buffers you have?

**Comment by YuriSolovyov** [comment]
I wonder if we can use [requestIdleCallback](https://developers.google.com/web/updates/2015/08/using-requestidlecallback) here once we have chrome 47+

**Comment by maxbrunsfeld** [comment]
> I wonder if we can use requestIdleCallback here once we have chrome 47+

Yeah, we should start using that, as soon as https://github.com/atom/atom/pull/9627 lands

File: src/storage-folder.coffee [hunk 3]
```diff
@@ -6,11 +6,16 @@ class StorageFolder
   constructor: (containingPath) ->
     @path = path.join(containingPath, "storage") if containingPath?
 
-  store: (name, object) ->
+  storeSync: (name, object) ->
     return unless @path?
 
     fs.writeFileSync(@pathForKey(name), JSON.stringify(object), 'utf8')
 
+  storeAsync: (name, object) ->
```
Comment (nathansobo) [inline src/storage-folder.coffee]: I think this should be called `store`... the convention in Node is to have async be implied and only mention sync vs async on a sync alternative.

**Comment by kuychaco** [comment]
Did some profiling. Here are durations for saveState (synchronous serialization, and async writing to disk):

| State | Time |
| ------- | ------- |
| 20 tabs opened | 2 ms |
| large unsaved buffer (jquery, 10K lines)* | 11 ms |

*Marker Layer serialization took 8.3 ms

20 tabs open:
screen shot 2016-02-01 at 5 58 09 pm

large unsaved buffer (jquery, 10K lines)
screen shot 2016-02-01 at 5 16 20 pm

**Comment by as-cii** [comment]
Great job, @kuychaco. I think those are really impressive results: the ideal idle time is `50ms`, which, if I am interpreting the results correctly, means we have room for serializing ~ 5 large files or ~ 400 open tabs.

Also, I find very interesting that most of the time is spent in dumping markers: I guess it's because we have to perform a full traversal of the tree, which can be time consuming for large files. I still wonder, though, if we can dissect that `MarkerIndex` call and maybe optimize based on that.

Also, @kuychaco, what's the time spent in `JSON.stringify`?

I am really excited about this! :sparkles:

**Comment by nathansobo** [comment]
> Also, I find very interesting that most of the time is spent in dumping markers

I'd like to implement serialization of the `MarkerIndex` directly for this case. Then we don't have to traverse the tree and build an alternate structure. Instead we can just convert the tree to JSON directly.

**Comment by kuychaco** [comment]
/cc @atom/core feedback welcome. Note that the travis failure can be disregarded ([flakey spec](https://github.com/atom/atom/issues/10435))

This PR is intended to improve crash recovery by serializing and saving the AtomEnvironment state regularly to minimize data loss.

**Serializing**
Serializing will occur upon debounced mousedown and keydown events (debouce interval is set to 1 second). As part of this PR, I've cleaned up the `AtomEnvironment` code and done away with the `state` instance variable. Assuming that the serialized state is successfully saved, that means that worst-case-scenario a lovely Atom user will lose only 1 second worth of work upon crash. And speaking of successful saving...

**Saving**
With this PR, the state is now stored in IndexedDB rather than the ~/.atom/storage folder. Here are some of the benefits of using IndexedDB:
- transactional nature means:
  - when the disk is full we won't lose data as we would in the case of truncated files on disk
  - when two Atom windows have the same project folders open we don't have to worry about race conditions
- IndexedDB can store JS objects so there is no need to JSON.stringify/parse the state (possible performance improvements for startup time if deserializing large state object)
- state can be easily accessed in chrome dev tools, which is pretty cool I think
 screen shot 2016-02-11 at 5 36 13 pm

TODO: come up with a user-friendly way to clear the state from IndexedDB for [debugging purposes](https://github.com/atom/atom/issues/2957)

File: src/atom-environment.coffee [hunk 4]
```diff
@@ -640,7 +648,7 @@ class AtomEnvironment extends Model
     @registerDefaultTargetForKeymaps()
 
     @packages.loadPackages()
-
+
```
Comment (nathansobo) [inline src/atom-environment.coffee]: :fire: this trailing whitespace

File: src/storage-folder.coffee [hunk 5]
```diff
@@ -6,11 +6,16 @@ class StorageFolder
   constructor: (containingPath) ->
     @path = path.join(containingPath, "storage") if containingPath?
 
-  store: (name, object) ->
+  storeSync: (name, object) ->
     return unless @path?
 
     fs.writeFileSync(@pathForKey(name), JSON.stringify(object), 'utf8')
 
+  store: (name, object) ->
```
Comment (nathansobo) [inline src/storage-folder.coffee]: Is this async version used now that you're using IndexedDB?

**Comment by nathansobo** [comment]
Very solid work @kuychaco. I did a pretty quick scan and things look good to me. Are you thinking of adding the `--clear-workspace-state` flag before merging?

**Comment by kuychaco** [comment]
@nathansobo thanks for the review! I think I will add the command-line flag before merging. Hope to get that done tonight.

**Comment by benogle** [comment]
:boom:

**Comment by jeancroy** [comment]
Hi @kuychaco 

Is there mechanism to migrate state data from previous format to the new IndexedDB format ?
Here's one thread where a beta user think he lost all his state data on upgrade, but it was there, just ignored/not loaded.

https://discuss.atom.io/t/upgrading-to-1-7-0-beta0-blew-away-my-session-any-way-to-restore/26918
It could be interesting to avoid a widespread fear of data loss.

**Comment by iolsen** [comment]
cc @as-cii ☝️

**Comment by nathansobo** [comment]
Talked with @as-cii and he's planning on falling back to reading the state out of `~/.atom` if it is present and there's nothing in the database.

**Comment by mehcode** [comment]
Great to hear there is a fallback in the works. :+1: 

---

> [...] People care more about their session state than I realized.

People (me) have even written a [package](https://github.com/mehcode/atom-project-plus) to list / inspect / make use of session state. It really is a nice feature of Atom.

**Review by miguelbarrio — COMMENTED** [review]
8d55bbcdeadc41b07a1e5e710e45b4a939c399eb
